snakePipes/shared/rules/filter_annotation.snakefile

```diff
@@ -1,54 +1,49 @@
+##     cols 13-15:    transcript_id, transcript_name, transcript_biotype
+##     cols 16-18:    gene_id, gene_name, gene_biotype
+##
+## and some gencode specific tags if present:
+##     cols 19+20:    gencode [basic, full, NA], basic=representative transcripts for each gene
+##     cols 21+22:    transcript_support_level [NA,1,2,3,4,5] 
+##        1 (all splice junctions of the transcript are supported by at least one non-suspect mRNA),
+##        2 (the best supporting mRNA is flagged as suspect or the support is from multiple ESTs),
+##        3 (the only support is from a single EST),
+##        4 (the best supporting EST is flagged as suspect),
+##        5 (no single transcript supports the model structure),
+##        NA (the transcript was not analyzed)  
+##     cols 23+24:    level [1,2,3,NA]  1=verified loci, 2=manually annotated loci, 3=automatically annotated loci
+##
+## according to the gtf format description, gene_id/transcript_id tags are required (by UCSCStools)
+## missing tags we set to NA or to the gene_id/transcript_id 
+rule create_annotation_bed:
+    input:
+        gtf = genes_gtf
+    output:
+        bed_annot = "Annotation/genes.annotated.bed"
+    conda: CONDA_RNASEQ_ENV
+    shell:
+        "join -t $'\t' -o auto --check-order -1 4 -2 1 "
+        "<(gtfToGenePred -ignoreGroupsWithoutExons {input.gtf} /dev/stdout | genePredToBed /dev/stdin /dev/stdout | tr ' ' $'\t' | sort -k4,4) "
+        "<(cat {input.gtf} | "
+        """ awk '$1!~"^#" && $3~"transcript|exon|start_codon|stop_codon|CDS"{{print $0}}' | tr -d "\\";" | """
+        """ awk '{{pos=match($0,"gene_id.([^[:space:]]+)",a); if (pos!=0) gid=a[1];else exit 1; """
+        """ pos=match($0,"transcript_id.([^[:space:]]+)",a); if (pos!=0) tid=a[1]; else tid=gid; """
+        """ pos=match($0,"transcript_name.([^[:space:]]+)",a); if (pos!=0) tna=a[1]; else tna=tid; """
+        """ pos=match($0,"transcript_[bio]*type.([^[:space:]]+)",a); if (pos!=0) tt=a[1]; else tt="NA"; """
+        """ pos=match($0,"gene_name.([^[:space:]]+)",a);  if (pos!=0) gna=a[1]; else gna=gid; """
+        """ pos=match($0,"gene_[bio]*type.([^[:space:]]+)",a); if (pos!=0) gt=a[1]; else gt="NA"; """
+        """ pos=match($0,"(transcript_support_level.[^[:space:]]+)",a); if (pos!=0) tsl=a[1]; else tsl="transcript_support_level NA"; """
+        """ pos=match($0,"[[:space:]](level.[^[:space:]]+)",a); if (pos!=0) lvl=a[1] ; else lvl="level NA"; """
+        """ pos=match($0,"tag.basic"); if (lvl!~"NA"){{if (pos==0) basic="full"; else basic="basic"}} else basic="NA"; """
+        """ OFS="\\t"; print tid,tna,tt,gid,gna,gt,"gencode",basic,"transcript_support_level",tsl,"level",lvl}}' | """
+        """ sort | uniq | sort -k1,1) | """
+        """ awk '{{OFS="\\t";print $2,$3,$4,$1,$5,$6,$7,$8,$9,$10,$11,$12,$1,$13,$14,$15,$16,$17,$18,$19,$20,$21,$22,$23}}' > {output.bed_annot} """ 
```

Comment:
I'd still prefer this to be a more readable python script long-term, but that can be saved for another day.

Reply:
yeah we can do this at some point, I tried to simplify the pipe as much as possible right now